### DISCUSSION TITLE
One to many, many to many relationship in one query

### DISCUSSION THREAD

**Discussion by JadKHaddad** [SeaQL/sea-orm#2106] [Q&A]
I have these simple Tables

users

| id | username |
| -- | -------- |

posts

| id | title | content | user_id |
| -- | ----- | ------- | ------- |

tags

| id | name |
| -- | ---- |

post_tags

| post_id | tag_id |
| ------- | ------ |

I want to get all users with their posts and tags.

```sql
SELECT users.*, ARRAY_AGG (posts.*) as posts
FROM (
        SELECT posts.*, ARRAY_AGG (tags.*) as tags
        FROM posts
            LEFT JOIN posts_tags ON posts_tags.post_id = posts.id
            LEFT JOIN tags ON tags.id = posts_tags.tag_id
        GROUP BY
            posts.id
    ) as posts
    -- Use LEFT JOIN to get only users that have posts
    RIGHT JOIN users ON users.id = posts.user_id
GROUP BY
    users.id;
```

Getting only the users with their posts is relatively easy using ```find_with_related```

Getting only the posts with their tags is also relatively easy using ```find_with_related```

But the problem is getting the users with their posts and tags. Currently, I have 2 ways to achieve this:

- Using ```find_with_related``` sequentially

```rust
// One to many
let users_with_posts = users::Entity::find()
    .find_with_related(posts::Entity)
    .all(db)
    .await?;

// Many to many
let posts_with_tags = posts::Entity::find()
    .find_with_related(tags::Entity)
    .all(db)
    .await?;

// Some iteration to merge the users_with_posts and posts_with_tags
// ---
```

- Using ```sea_query``` to write a custom query, which looks wild by the way.

```rust
let query = sea_query::Query::select()
        .column((users::Entity, users::Column::Id))
        .column((users::Entity, users::Column::Username))
        .expr(Expr::cust(
            "ARRAY_AGG (
                row_to_json (posts.*)
            ) as posts",
        ))
        .from_subquery(
            sea_query::Query::select()
                .column((posts::Entity, posts::Column::Id))
                .column((posts::Entity, posts::Column::Title))
                .column((posts::Entity, posts::Column::Content))
                .column((posts::Entity, posts::Column::UserId))
                .expr(Expr::cust(
                    "ARRAY_AGG (
                        row_to_json (tags.*)
                    ) as tags",
                ))
                .from(posts::Entity)
                .left_join(
                    posts_tags::Entity,
                    Expr::col((posts::Entity, posts::Column::Id)).equals((
                        posts_tags::Entity,
                        posts_tags::Column::PostId,
                    )),
                )
                .left_join(
                    tags::Entity,
                    Expr::col((
                        posts_tags::Entity,
                        posts_tags::Column::TagId,
                    ))
                    .equals((tags::Entity, tags::Column::Id)),
                )
                .group_by_col((posts::Entity, posts::Column::Id))
                .take(),
            sea_query::Alias::new("posts"),
        )
        // Use LEFT JOIN to get only users that have posts
        .right_join(
            users::Entity,
            Expr::col((users::Entity, users::Column::Id))
                .equals((posts::Entity, posts::Column::UserId)),
        )
        .group_by_col((users::Entity, users::Column::Id))
        .to_owned();

    let builder = db.get_database_backend();
    let stmt = builder.build(&query);

    let res = db.query_all(stmt).await?;
    let res = res
        .into_iter()
        .map(|q_res| {
            let json = sea_orm::JsonValue::from_query_result(&q_res, "").unwrap_or_default();
            
            // UersPostTags is a custom type that implements Deserialize
            let user_posts_tags: Option<UersPostTags> =
                serde_json::from_value(json).unwrap_or_default();

            println!("{:#?}", user_posts_tags);

            user_posts_tags
        })
        .collect::<Vec<_>>();
```

which is equivalent to:

```sql
SELECT users.*, ARRAY_AGG (row_to_json (posts.*)) as posts
FROM (
        SELECT posts.*, ARRAY_AGG (row_to_json (tags.*)) as tags
        FROM posts
            LEFT JOIN posts_tags ON posts_tags.post_id = posts.id
            LEFT JOIN tags ON tags.id = posts_tags.tag_id
        GROUP BY
            posts.id
    ) as posts
    -- Use LEFT JOIN to get only users that have posts
    RIGHT JOIN users ON users.id = posts.user_id
GROUP BY
    users.id;
```

Notice that I'm also using ```row_to_json``` to convert the rows to ```JsonValue```.

So basically, I want to perform ```one-to-many```, ```many-to-many``` relationship query using ```sea-orm```. Is there a better way to achieve this? Thanks in advance.
